### ISSUE TITLE
C++ Core Guidelines should introduce more smart pointers

### ISSUE THREAD

**Issue by NN---** [isocpp/CppCoreGuidelines#1102]
C++ Core Guidelines suggest to use either unique_ptr or shared_ptr for better code.

The main problem that there are cases where unique_ptr is better to be replaced by other type of smart pointers.
Many developers don't consider this option at all and introduce bugs in the code due to lack of appropriate semantic for the smart pointer.

Examples of pointers that can be introduced.

```cpp
void f()
{
 // unique_ptr allows changing the object while scoped_ptr just provides exception safety and nothing more
 // unique_ptr<Object> o(new Object());   
 // local_ptr is non-copyable, non-movable and non-resettable object
 local_ptr<Object> o(new Object()); 
}

void g()
{
 // non-copyable, non-movable but can be reset
 scoped_ptr<Object> o(new Object()); 
 
 // OK
 o.reset(new Object());
}

class A
{
public:
 A() : o(new Object()) {}

private:
 // Again, unique_ptr allows object to be changed and moved away while local_ptr is not movable nor copyable.
 local_ptr<Object> o;
};

class A
{
public:
 A() {}
 void f()
 {
   o.reset(new Object());
   // Calling reset again is an error.
 }

private:
 // Allows initialization only once
 initialization_ptr<Object> o;
};
```

My suggestion is to cover the popular use-cases with semantic driven smart pointers.
This allows analyzers to understand what is the developer intention and also prevents bugs in the code.

**Comment by jwakely** [comment]
Can't half your examples use `const unique_ptr<Object>`?

**Comment by NN---** [comment]
@jwakely Good point.
I will revisit examples and update the proposal with const unique_ptr.

**Comment by cubbimew** [comment]
>     const unique_ptr<Object> o(new Object()); 

following [R.11 : Avoid calling new and delete explicitly](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rr-newdelete), how about

      const auto p = std::make_unique<Object>();

**Comment by AndrewPardoe** [comment]
@NN--- Thank you for this suggestion! Could you please prepare a PR with your changes?

**Comment by NN---** [comment]
@cubbimew Interesting point.
On the other hand there is class template argument deduction which reduce need of make_* functions.
http://en.cppreference.com/w/cpp/language/class_template_argument_deduction 
Seems like there is something missing which will be both safe and with type deduction.

**Comment by blakehawkins** [comment]
With `make_*` functions you can avoid using the `new` operator in product code.

Also, class-template argument deduction doesn't apply to c++14.

**Comment by jwakely** [comment]
> With `make_*` functions you can avoid using the new operator in product code.

That's true for `make_unique` and `make_shared` but they aren't relevant for class template argument deduction. I assume the previous comment was talking about [object generators](http://www.boost.org/community/generic_programming.html#object_generator) such as `make_pair`, `make_tuple`, `make_optional` etc. which exists to deduce the result type from the arguments. You wouldn't be using `new` in those cases anyway.
